WIP: Feat/distributions

seidlj opened the pull request:
This PR adds module `distributions` which treats the virtual samples as a representation of the distribution itself. Now the user can do something like:
```
gamma_distr = distributions.gamma(a=5, loc=5)
samples = sample(gamma_distr, 1000)
more_samples = sample(gamma_distr, 100)
```

To sample the distribution right away one can use `sample=True`: 
```
distributions.normal(scale=5, samples=100, sample=True) # -> 100 samples
```

This feels quite intuitive to me. @smartass101, what do you think? If you agree with this approach, I'll add shortcuts to more distributions before merging so that the user does not need to call `distributions.from_scipy` so often.

smartass101 commented:
First, spontaneous reactions:
- nice!
- `samples=100, sample=True)` does not seem very intuitive to me, perhaps by default we could have `samples=VIRTUAL_SAMPLE` with a placeholder `VIRTUAL_SAMPLE = -1`
- instead of making the ditribution generator functions as module-level functions, we could create a `DistributionsFactory` object available under `distributions` which could dynamically create them on attribute access, tab-completion can be facilitated by the `AttrMixing` form xarray

seidlj committed:
add all scipy distributions, make api closer to scipy

seidlj committed:
rename `distributions` module to `stats`

seidlj committed:
move function `sample` to module `sampling`

seidlj committed:
move virtually sampled distributions into module `distributions`

seidlj committed:
bind the rest of scipy rv methods (still return numpy arrays)

seidlj committed:
Implement frozen distributions

smartass101 commented on xrrandom/stats.py:
```diff
@@ -0,0 +1,163 @@
+                            'mean', 'std', 'var', 'interval'}
+
+def _bind_stats_method(dest, source, method):
+    return getattr(source, method).__get__(dest)
```
cool but cryptic... Using `MethodType(...)` would be more readable and explicit

seidlj replied:
This piece of code is no longer used. The methods are now added to a class.

smartass101 commented on xrrandom/stats.py:
```diff
@@ -0,0 +1,163 @@
+        self._distr = distr
+        self.__doc__ = distr.__doc__.split('\n')[0]        
+        for method in _common_scipy_rv_methods | _scipy_rv_methods[distribution_kind(distr)]:
+            setattr(self, method, _bind_stats_method(self, distr, method))    
```
we could instead create two classes (via mixins with common functions) which would define the wrappers at the class level, skipping repeated calls during each initialization.

seidlj replied:
In order to be able to modify doc string of dunder methods, I had to create a separate class for each distribution. This part is now called only once during module import.

smartass101 commented on xrrandom/stats.py:
```diff
@@ -0,0 +1,163 @@
+        self.virtual = virtual
+
+        for method in _common_scipy_rv_methods | _scipy_rv_methods[distribution_kind(distr._distr)]:
+            setattr(self, method, _bind_frozen_method(self, distr, method))
```
Could also be done via a mixin at the class level similar to the above, just the methods would access `self.args` and `self.kwargs`

smartass101 commented on xrrandom/stats.py:
```diff
@@ -0,0 +1,163 @@
+        return self.distr.rvs(*self.args, samples=samples, virtual=virtual, sample_chunksizes=sample_chunksize, **self.kwargs)
+
+# augment this module by all distributions in the scipy.stats
+for name, distr in stats.__dict__.items():
```
using `vars()` might be more future proof and backwards compatible if the `__dict__` interface changes.

seidlj replied:
Done.

smartass101 commented on xrrandom/distributions.py:
```diff
@@ -0,0 +1,55 @@
+    """   
+
+    def __init__(self, distr):
+        self._distr = distr
```
could be used in `__repr__`

seidlj replied:
I've added `__repr__` for frozen distributions, the default representation seems ok for the other classes.

smartass101 reviewed (commented):
Nice concept. I would suggest to use Mixins instead of repeatedly setting the same methods for each instance. But if you at some point want to set a distribution-dependent signature for each instance, then this indeed would be the way.

seidlj committed:
Change signature of the distribution methods to show shape parameters

seidlj committed:
Rewrite frozen distributions

seidlj committed:
Create unique class for each scipy distribution 

Adding methods to distribution class (instead of instance) allows setting doc strings for dunder methods such as __call__. Moreover, the added methods are now visible in help.

seidlj committed:
Fix issue with non-shape parameters not beeing passed to scipy

seidlj committed:
Return DataArray

seidlj committed:
Add docstring to frozen distributions.

seidlj committed:
Show shape parameters in doc strings

seidlj committed:
Replace `__dict__` with `vars`

seidlj committed:
Describe new features in readme

seidlj committed:
add basics of distributions module

add normal and gamma distributions
import `sample` function in the package init

seidlj committed:
Nice repr for frozen distributions

seidlj committed:
Raise TypeError if some shape parameters are missing

seidlj committed:
Add tests

fixup! Add tests
